src/uthash.h

```diff
@@ -999,6 +999,38 @@ for(((el)=(head)), ((tmp)=DECLTYPE(el)((head!=NULL)?(head)->hh.next:NULL));
   (el) != NULL; ((el)=(tmp)), ((tmp)=DECLTYPE(el)((tmp!=NULL)?(tmp)->hh.next:NULL)))
 #endif
 
+#ifdef NO_DECLTYPE
+#define HASH_TOP(hh,head,el,tmp,eof)                                             \
+do {                                                                             \
+  ((el)=(head));                                                                 \
+  ((*(char**)(&(tmp)))=(char*)((head!=NULL)?(head)->hh.next:NULL));              \
+  ((*eof)=((el) == NULL));                                                       \
+} while(0)
+#else
+#define HASH_TOP(hh,head,el,tmp,eof)                                             \
+do {                                                                             \
+  ((el)=(head));                                                                 \
+  ((tmp)=DECLTYPE(el)((head!=NULL)?(head)->hh.next:NULL));                       \
+  ((*eof)=((el) == NULL));                                                       \
+} while(0)
+#endif
+
+#ifdef NO_DECLTYPE
+#define HASH_NEXT(hh,head,el,tmp,eof)                                            \
+do {                                                                             \
+  ((el)=(tmp));                                                                  \
+  ((*(char**)(&(tmp)))=(char*)((tmp!=NULL)?(tmp)->hh.next:NULL));                \
+  ((*eof)=((el) == NULL));                                                       \
```

Comment:
If I understand correctly, you've got three pieces of state here:
- `el` (the pointer to the actual current element)
- `tmp` (the pointer to the next element, must never be touched by anyone except these macros)
- `eof` (a bool that's invariably equal to `!el`)

I'd say that `eof` should just disappear.
It would be *nice* if `tmp` and `el` could be rolled up together into a single `struct UT_hash_iterator` or something, but I guess that's overkill in that it wouldn't actually make the code any easier to use. (The user would have to reach into the struct to pull out `el`, and pulling out `tmp` instead would still be catastrophic.)